geode-management/src/main/java/org/apache/geode/management/client/ClusterManagementServiceProvider.java

```diff
@@ -0,0 +1,135 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+import org.springframework.http.client.ClientHttpRequestFactory;
+
+import org.apache.geode.annotations.Experimental;
+import org.apache.geode.management.api.ClusterManagementService;
+import org.apache.geode.management.spi.ClusterManagementServiceProviderFactory;
+
+/**
+ * Top-level entry point for client interaction with the {@link ClusterManagementService}. A user
+ * can create an instance of the {@code ClusterManagementService} (CMS for short) in several ways,
+ * each providing various level of control or expediency.
+ * <p/>
+ * Calling {@code getFactory(context)} will return an explicit instance of {@link
+ * ClusterManagementServiceProviderFactory}. Methods on this factory can then be called to create or
+ * retrieve instances of a CMS. New {@link ClusterManagementServiceProvider}s can be written if
+ * specific customization or parameterization is required.
+ * <p/>
+ * A note about contexts. A context is simply a unique string which identifies a specific instance
+ * of CMS that will be returned. Contexts map to the different uses of a CMS. Currently, the
+ * following contexts are provided:
+ * <ul>
+ * <li>
+ * {@code JAVA_CLIENT_CONTEXT} ("java-client") - would be used to retrieve a CMS instance on a pure
+ * Java client - i.e. an app that is not running in the context of a Geode client or server. This
+ * context is available when using the <i>geode-management</i> module.
+ * </li>
+ * <li>{@code GEODE_CLIENT_CONTEXT} ("geode-client") - would be used to retrieve a CMS instance
+ * from a JVM where a {@code ClientCache} exists. This context is available when using the
+ * <i>geode-core</i> module.
+ * </li>
+ * <li>
+ * {@code LOCATOR_CONTEXT} ("locator") - would be used to retrieve a CMS instance from a JVM which
+ * is hosting a locator. This context should typically only be used for internal Geode code and is
+ * not particularly relevant to client or user applications.
+ * </li>
+ * <li>
+ * {@code SERVER_CONTEXT} ("server") - would be used to retrieve a CMS instance from a JVM where a
+ * {@code Cache} exists. This would be either a server or locator. This context is available when
+ * using the <i>geode-core</i> module.
+ * </li>
+ * </ul>
+ * If the URL of the Cluster Management Service is known, the {@code getService(url)} method can be
+ * called. This implicitly uses the {@code JAVA_CLIENT_CONTEXT}. For further control the CMS can be
+ * configured with a {@link ClientHttpRequestFactory} by calling {@code
+ * getService(requestFactory)}.
+ * <p/>
+ * Finally, the simplest way to create a CMS instance is simply to call {@code getService()}. This
+ * method
+ * will attempt to infer the context and use an appropriate service provider to create a CMS
+ * instance.
+ */
+@Experimental
+public class ClusterManagementServiceProvider {
+
+  public static final String JAVA_CLIENT_CONTEXT = "java-client";
```

Comment:
The way I understand is, for developer facing, there are two ways to get the CMS, one is without the url (in the server context), one is with the url (in the client context, either with a url or with an HttpRequestFactory). Can we put all the service create logic in this `ClusterManagementServiceProvider` class instead of going through all the factories? i.e, your getService() method body is the `SmartClusterManagementServiceProviderFactory` create, and getService(url) is the "BasicClusterManagementServiceProviderFactory` create. This seems much easier to follow and without all these service loader stuff.